multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/cf/CloudControllerClientProvider.java

```diff
@@ -79,116 +54,29 @@ public CloudControllerClient getControllerClientWithNoCorrelation(String userNam
-     * @param correlationId of the process which is used to tag HTTP requests
-     */
-    public void releaseClient(String userName, String spaceGuid, String correlationId) {
-        clients.remove(getKey(userName, spaceGuid, correlationId));
-    }
-
     /**
      * Releases the client for the specified user name and space id by removing it from the clients cache.
      *
      * @param userName the user name associated with the client
      * @param spaceGuid the space id associated with the client
      */
-    public void releaseClientWithNoCorrelation(String userName, String spaceGuid) {
-        clients.remove(getKey(userName, spaceGuid, NO_CORRELATION_ID));
-    }
-
-    /**
-     * Releases the client for the specified user name, organization and space by removing it from the clients cache.
-     *
-     * @param userName the user name associated with the client
-     * @param org the organization associated with the client
-     * @param space the space associated with the client
-     */
-    public void releaseClientWithNoCorrelation(String userName, String org, String space) {
-        clients.remove(getKey(userName, org, space, NO_CORRELATION_ID));
-    }
-
-    private OAuth2AccessTokenWithAdditionalInfo getValidToken(String userName) {
-        OAuth2AccessTokenWithAdditionalInfo token = tokenService.getToken(userName);
-        if (token.getOAuth2AccessToken()
-                 .getExpiresAt()
-                 .isBefore(Instant.now())) {
-            throw new SLException(Messages.TOKEN_EXPIRED, userName);
-        }
-        return token;
-    }
-
-    private CloudControllerClient getClientFromCacheWithNoCorrelation(String userName, String org, String space) {
-        String key = getKey(userName, org, space, NO_CORRELATION_ID);
-        return clients.computeIfAbsent(key,
-                                       k -> clientFactory.createClient(getValidToken(userName), org, space, null));
-    }
-
-    private CloudControllerClient getClientFromCache(String userName, String org, String space, String correlationId) {
-        String key = getKey(userName, org, space, correlationId);
-        return clients.computeIfAbsent(key,
-                                       k -> clientFactory.createClient(getValidToken(userName), org, space, correlationId));
+    public void releaseClient(String userName, String spaceGuid) {
+        clients.remove(getKey(userName, spaceGuid));
     }
 
     private CloudControllerClient getClientFromCacheWithNoCorrelation(String userName, String spaceId) {
-        String key = getKey(userName, spaceId, NO_CORRELATION_ID);
+        String key = getKey(spaceId, userName);
         return clients.computeIfAbsent(key,
-                                       k -> clientFactory.createClient(getValidToken(userName), spaceId, null));
+                                       () -> clientFactory.createClient(tokenService.getToken(userName), spaceId, null));
```

Comment:
Yes, it's used.
The flow for the tokens goes like this:
```
- get token from cache
- create oauth client object (contains a web client, a token factory, last used token and others..)
- create ClientCredentials object with the retrieved token
- initialize the oauth client with those credentials (sets the token field in OAuthClient)
- create the cloud controller client
```
Then, on calls to CF:
```
- OAuthClient.getToken() 
- check if the token field is set
   - if set and the token expires after more than 50 seconds, return it
   - otherwise, make a call to the UAA endpoint to fetch a new one and update the field
- add token to authorization header
```

Reply:
Ok, I agree it is used. Do you think it makes sense (with another backlog) to introduce a unified way of getting tokens? Currently, we use:
oAuthClient.getToken() - For Oauth and Basic depending if an extended client is created or not.
oAuthClient.getAuthorizationHeaderValue() - to get the token value

What seems the best in my opinion is some sort of a filter as in "CloudControllerRestClientRequestFilterFunction", right before every call to CC to renew the token. The custom clients have to be changed as well as the logging client.
There are a lot of places where we invoke oAuthClient.init which might be missed.